### PR TITLE
Test triggers in scenarios

### PR DESCRIPTION
**Pull request by aherrmann-da**
This allows testing a trigger's `rule` in a `scenario` by providing the `testRule` function. It takes the `Party`, the ACS, the commands in flight, and the trigger state as parameters and returns the `[Commands]` issued by the trigger.

The user provides the ACS using the `ACSBuilder` type. This way it is sufficient for the user to provide `ContractId t`s and the fetching of the contract value and the conversion to `AnyContractId`/`AnyTemplate` is handled automatically.

The user can perform assertions on the emitted commands using `assertCreateCmd`, `assertExerciseCmd`, and `assertExerciseByKeyCmd`. These take a list of emitted `Command`s and the assertions as `a -> Either Text ()`. If none of the commands fulfills the assertions, then the scenario will fail.

To facilitate the above, this PR adds an `instance CanAbort (Either Text)`. `Either` already had the instances `Action`, and `ActionFail` before.

This PR adds a simple test case for `testRule` and the assertion functions.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-stdlib-src/DA/Internal/LF.daml [hunk 1]
```diff
@@ -179,6 +179,9 @@ instance CanAbort Update where
 instance CanAbort Scenario where
     abort = fail
 
+instance CanAbort (Either Text) where
```
Comment (cocreature) [inline compiler/damlc/daml-stdlib-src/DA/Internal/LF.daml]: :+1:

File: triggers/daml/Daml/Trigger.daml [hunk 2]
```diff
@@ -28,13 +28,21 @@ module Daml.Trigger
  , RegisteredTemplates(..)
  , registeredTemplate
  , RelTime(..)
+ , toACS
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger.daml]: Let’s leave it here for this PR but I wonder if it might make sense to move the testing functions to something like `Daml.Trigger.Test` or `Daml.Trigger.Assert` to match `DA.Assert`.

Comment (hurryabit) [inline triggers/daml/Daml/Trigger.daml]: Shouldn't we export the `ACSBuilder` _type_ as well?

File: triggers/daml/Daml/Trigger.daml [hunk 3]
```diff
@@ -260,3 +268,109 @@ data TriggerState s = TriggerState
   , commandsInFlight : Map CommandId [Command]
   , nextCommandId : Int
   }
+
+-- | Used to construct an 'ACS' for 'testRule'.
+newtype ACSBuilder = ACSBuilder [Update (AnyContractId, AnyTemplate)]
```
Comment (hurryabit) [inline triggers/daml/Daml/Trigger.daml]: `GeneralizedNewtypeDeriving` should allows us to write
```suggestion
newtype ACSBuilder = ACSBuilder [Update (AnyContractId, AnyTemplate)]
  deriving (Semigroup, Monoid)
```

File: triggers/daml/Daml/Trigger.daml [hunk 4]
```diff
@@ -260,3 +268,109 @@ data TriggerState s = TriggerState
...
+  where
+    step : Either [Text] () -> Command -> Either [Text] ()
+    step (Right ()) _ = Right ()
+    step (Left msgs) command =
+      case assertion <$> fromCommand command of
+        None -> Left msgs
+        Some (Left msg) -> Left (msg :: msgs)
+        Some (Right ()) -> Right ()
+
+-- | Check that at least one command fulfills the given assertions.
```
Comment (hurryabit) [inline triggers/daml/Daml/Trigger.daml]:
```suggestion
-- | Check that at least one command is a create command whose payload fulfills the given assertions.
```

File: triggers/daml/Daml/Trigger.daml [hunk 5]
```diff
@@ -260,3 +268,109 @@ data TriggerState s = TriggerState
...
+  => [Command]  -- ^ Check these commands.
+  -> (t -> Either Text ())  -- ^ Perform these assertions.
+  -> m ()
+assertCreateCmd commands assertion =
+  case expectCommand commands fromCreate assertion of
+    Right () -> pure ()
+    Left msgs ->
+      abort $ "Failure, found no matching create command." <> collectMessages msgs
+
+-- | Check that at least one command fulfills the given assertions.
```
Comment (hurryabit) [inline triggers/daml/Daml/Trigger.daml]:
```suggestion
-- | Check that at least one command is an exercise command whose contract id and choice argument fulfill the given assertions.
```
Similarly for `assertExerciseByKeyCmd`.

File: triggers/tests/scenarios/Rule.daml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      -- Exercise a choice by key
+      dedupExerciseByKey @T (party, 0) Poke with n = 1
+  registeredTemplates = RegisteredTemplates [registeredTemplate @T]
+  heartbeat = None
+
+test = scenario do
+  alice <- getParty "Alice"
+  tId <- submit alice do create T with party = alice, count = 1
+  let activeContracts = toACS tId
+      commandsInFlight = Map.empty
```
Comment (hurryabit) [inline triggers/tests/scenarios/Rule.daml]:
```suggestion
  let commandsInFlight = Map.empty
```
Better safe than sorry. 😃

File: triggers/daml/Daml/Trigger.daml [hunk 7]
```diff
@@ -260,3 +268,109 @@ data TriggerState s = TriggerState
   , commandsInFlight : Map CommandId [Command]
   , nextCommandId : Int
   }
+
+-- | Used to construct an 'ACS' for 'testRule'.
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger.daml]: We should probably export this type but without the constructor. We leak it anyway since we expose `toACS`.

File: triggers/daml/Daml/Trigger.daml (hunk 2 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: Indeed, done.

File: triggers/daml/Daml/Trigger.daml [hunk 8]
```diff
@@ -260,3 +268,109 @@ data TriggerState s = TriggerState
...
+    step : Either [Text] () -> Command -> Either [Text] ()
+    step (Right ()) _ = Right ()
+    step (Left msgs) command =
+      case assertion <$> fromCommand command of
+        None -> Left msgs
+        Some (Left msg) -> Left (msg :: msgs)
+        Some (Right ()) -> Right ()
+
+-- | Check that at least one command fulfills the given assertions.
+assertCreateCmd
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger.daml]: With these assertion helpers it is easy to check that a given command was emitted but it is tricky to assert that a command was not emitted or that a given command was emitted multiple times (although I don’t really see a good usecase for the latter). Not sure how important this is so let’s definitely leave it for a separate PR but worth keeping in mind.

File: triggers/daml/Daml/Trigger.daml (hunk 3 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: I tried that, but it fails with
```
Severity: DsError
Message:
  Failure to process DAML program, this feature is not currently supported.
  Coercion with tyconappco.
  ([Sym (N:ACSBuilder[0])])_R
```

File: triggers/daml/Daml/Trigger.daml (hunk 4 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: done.

File: triggers/daml/Daml/Trigger.daml (hunk 5 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: done

File: triggers/tests/scenarios/Rule.daml (hunk 6 above)
Comment (aherrmann-da) [reply triggers/tests/scenarios/Rule.daml]: done

File: triggers/daml/Daml/Trigger.daml (hunk 3 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger.daml]: Openedhttps://github.com/digital-asset/daml/issues/4237 to track this

File: triggers/daml/Daml/Trigger.daml (hunk 2 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: Yeah, that sounds like a good idea. In the specific case of `ACSBuilder`, we'd either have to also export `ACS (..)` from `Daml.Trigger`, or keep `ACSBuilder` and `buildACS` in `Daml.Trigger`.

File: triggers/daml/Daml/Trigger.daml (hunk 7 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: Yes, done.

File: triggers/daml/Daml/Trigger.daml (hunk 2 above)
Comment (cocreature) [reply triggers/daml/Daml/Trigger.daml]: We could have a separate `Daml.Trigger.Internal` module that exports `ACS(..)` and import it from both `Daml.Trigger` which then only needs to reexport `ACS` and `Daml.Trigger.Assert`.

File: triggers/daml/Daml/Trigger.daml (hunk 8 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: Yes, agreed. I'd say this is a first draft at the assertion API. We'll have to see what API is demanded as use-cases come up.

File: triggers/daml/Daml/Trigger.daml (hunk 3 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: Thanks!

File: triggers/daml/Daml/Trigger.daml (hunk 2 above)
Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger.daml]: Yeah, that sounds good.
